File: 3225-maximum-score-from-grid-operations/3225-maximum-score-from-grid-operations.java
class Solution {
  public long maximumScore(int[][] grid) {
    final int n = grid.length;
    long[][] prefix = new long[n][n + 1];
    long[] prevPick = new long[n + 1];
    long[] prevSkip = new long[n + 1];
    for (int j = 0; j < n; ++j)
      for (int i = 0; i < n; ++i)
        prefix[j][i + 1] = prefix[j][i] + grid[i][j];
    for (int j = 1; j < n; ++j) {
      long[] currPick = new long[n + 1];
      long[] currSkip = new long[n + 1];
      for (int curr = 0; curr <= n; ++curr)
        for (int prev = 0; prev <= n; ++prev)
          if (curr > prev) {
            final long score = prefix[j - 1][curr] - prefix[j - 1][prev];
            currPick[curr] = Math.max(currPick[curr], prevSkip[prev] + score);
            currSkip[curr] = Math.max(currSkip[curr], prevSkip[prev] + score);
          } else {
            final long score = prefix[j][prev] - prefix[j][curr];
            currPick[curr] = Math.max(currPick[curr], prevPick[prev] + score);
            currSkip[curr] = Math.max(currSkip[curr], prevPick[prev]);
          }
      prevPick = currPick;
      prevSkip = currSkip;
    }
    return Arrays.stream(prevPick).max().getAsLong();
  }
}

File: 3225-maximum-score-from-grid-operations/3225-maximum-score-from-grid-operations.py
class Solution:
  def maximumScore(self, grid: List[List[int]]) -> int:
    n = len(grid)
    p = [[0] * (n + 1) for _ in range(n)]
    a = [0] * (n + 1)
    b = [0] * (n + 1)
    for j in range(n):
      for i in range(n):
        p[j][i + 1] = p[j][i] + grid[i][j]
    for j in range(1, n):
      c = [0] * (n + 1)
      d = [0] * (n + 1)
      for x in range(n + 1):
        for y in range(n + 1):
          if x > y:
            s = p[j - 1][x] - p[j - 1][y]
            c[x] = max(c[x], b[y] + s)
            d[x] = max(d[x], b[y] + s)
          else:
            s = p[j][y] - p[j][x]
            c[x] = max(c[x], a[y] + s)
            d[x] = max(d[x], a[y])
      a = c
      b = d
    return max(a)
